Enforce rref python pickling to be in the scope of RPC call

zhaojuanmao opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#34755 Enforce rref python pickling to be in the scope of RPC call**

This diff disallows to use python pickler to pickle RRef. RRef can only be pickled in the scope of RPC call using _InternalRPCPickler.

Differential Revision: [D20453806](https://our.internmc.facebook.com/intern/diff/D20453806/)

zhaojuanmao committed:
Enforce rref python pickling to be in the scope of RPC call

This diff disallows to use python pickler to pickle RRef. RRef can only be pickled in the scope of RPC call using _InternalRPCPickler.

Differential Revision: [D20453806](https://our.internmc.facebook.com/intern/diff/D20453806/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Enforce rref python pickling to be in the scope of RPC call"


This diff disallows to use python pickler to pickle RRef. RRef can only be pickled in the scope of RPC call using _InternalRPCPickler.

Differential Revision: [D20453806](https://our.internmc.facebook.com/intern/diff/D20453806/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Enforce rref python pickling to be in the scope of RPC call"


This diff disallows to use python pickler to pickle RRef. RRef can only be pickled in the scope of RPC call using _InternalRPCPickler.

Differential Revision: [D20453806](https://our.internmc.facebook.com/intern/diff/D20453806/)

[ghstack-poisoned]

mrshenli commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -195,6 +195,9 @@ PyObject* rpc_init(PyObject* /* unused */) {
                   [](const PyRRef& self) {
+                    TORCH_CHECK(
+                        false,
+                        "Can not pickle rref in python pickler, rref can only be pickled in _InternalRPCPickler");
```
Since this will be a user message, does it make sense if we don't mention `_InternalRPCPickler` and only say "RRef can only be pickled when using RPC".

zhaojuanmao replied:
sounds good

mrshenli commented on torch/distributed/rpc/internal.py:
```diff
@@ -58,6 +67,14 @@ def serialize(self, obj):
+        # rpc api could accept user picklers inheriting from _InternalRPCPickler to serialize rref,
+        # user picklers could have different initialization function from _InternalRPCPickler,
+        # but all the user picklers should call serialize() and use _rref_reducer to pickle rref
+        # in python. also, when _internalRPCPickler is imported to rpc/api.py, rpc.RRef is not 
```
This is call `_internal_rpc_pickler` in the code

mrshenli replied:
Or do you mean the `_InternalRPCPickler ` class? We didn't import the class type in `rpc/api.py`, no?

zhaojuanmao replied:
will change

zhaojuanmao replied:
no, we did not import class type, just import the global pickler

mrshenli commented on torch/distributed/rpc/internal.py:
```diff
@@ -58,6 +67,14 @@ def serialize(self, obj):
+        # user picklers could have different initialization function from _InternalRPCPickler,
+        # but all the user picklers should call serialize() and use _rref_reducer to pickle rref
+        # in python. also, when _internalRPCPickler is imported to rpc/api.py, rpc.RRef is not 
+        # compiled yet, it is not good place to acces rpc.RRef inside _InternalRPCPickler constructor,
```
What determine the order of compilation? Does it work if we explicitly do `from distributed.rpc import RRef` in this file?

zhaojuanmao replied:
from .. import .. seems does not work, we can not import rpc in this file as rpc depends on this file, there is circular import issue.

mrshenli reviewed (commented):
LGTM! Just have some questions inline.

mrshenli commented on torch/distributed/rpc/internal.py:
```diff
@@ -49,6 +50,14 @@ def _tensor_reducer(self, obj):
+    def _rref_receiver(cls, rref_fork_data):
+        return dist.rpc.RRef._deserialize(rref_fork_data)
+
+    def _rref_reducer(self, obj):
```
We expect `obj` to be an RRef? Shall we explicitly call it rref here? 

Same for the `_tensor_reducer` above.

zhaojuanmao replied:
sounds good

zhaojuanmao committed:
Update on "Enforce rref python pickling to be in the scope of RPC call"


This diff disallows to use python pickler to pickle RRef. RRef can only be pickled in the scope of RPC call using _InternalRPCPickler.

Differential Revision: [D20453806](https://our.internmc.facebook.com/intern/diff/D20453806/)

[ghstack-poisoned]

pritamdamania87 commented on torch/csrc/distributed/rpc/init.cpp:
```diff
@@ -225,6 +228,14 @@ PyObject* rpc_init(PyObject* /* unused */) {
                     return PyRRef::unpickle(t);
```
Shouldn't we have an TORCH_CHECK here as well?

zhaojuanmao replied:
If rref can not be pickled, it can not be unpickled as well? but I think it is not harm to add one more check, will add it

zhaojuanmao committed:
Update on "Enforce rref python pickling to be in the scope of RPC call"


This diff disallows to use python pickler to pickle RRef. RRef can only be pickled in the scope of RPC call using _InternalRPCPickler.

Differential Revision: [D20453806](https://our.internmc.facebook.com/intern/diff/D20453806/)

[ghstack-poisoned]

zhaojuanmao committed:
Update on "Enforce rref python pickling to be in the scope of RPC call"


This diff disallows to use python pickler to pickle RRef. RRef can only be pickled in the scope of RPC call using _InternalRPCPickler.

Differential Revision: [D20453806](https://our.internmc.facebook.com/intern/diff/D20453806/)

[ghstack-poisoned]
